### ISSUE TITLE
[Span][FormattedString] triggers to many unwanted label text updates

### ISSUE THREAD

**Issue by farfromrefug** [NativeScript/NativeScript#8341]
<img width="1920" alt="Screen Shot 2020-02-14 at 13 22 07" src="https://user-images.githubusercontent.com/655344/74531940-4997da80-4f2e-11ea-809d-6b6b7333af31.png">
It is actually easy to understand 
* my label has color set in css
* my label has 6 spans
* `updateDynamicState` => triggers a colorProperty call on the label
* each span is notified of the property changed for label color
* each span then triggers a `FormattedString` on change event
* the label received 6 events for `FormattedString` changes => 6 setFormattedText

Actually i think it makes no sense for the Span / FormattedString to listen for those properties.
It is the job of the label to update itself

**Comment by edusperoni** [comment]
I've worked a lot with text on listviews for a project and also had this issue. In the end I used a directive that would create a new FormattedString and set it, triggering only once.

Handing text, especially of different sizes and formats, is hard.

Creating a new spannablestring is very expensive. Depending on the complexity of the spans it can take over 15ms for a single string, which means several dropped frames on scroll.

To efficiently show text you'd have to cache all native objects (hash bg color, color, and other layout affecting properties + formattedstring data and use it as key to the native object). This alone is not enough, as you need to calculate the text size, which is another headache. Instagram has great resources on it https://instagram-engineering.com/improving-comment-rendering-on-android-a77d5db3d82e

The idea is move all text calculating logic into a background process (I've tried with workers, as we can share native objects - I've used a native helper class, as we can't pass them as messages), ideally when you're requesting the data, as you're waiting for the network anyway. For text sizes we can use Android's own `setTextFuture` instead of `setText` (AppCompatTextView instead of TextView) which will calculate and cache the text sizes for us, so we don't need to calculate it ourselves: https://developer.android.com/reference/androidx/appcompat/widget/AppCompatTextView#setTextFuture(java.util.concurrent.Future%3Candroidx.core.text.PrecomputedTextCompat%3E)


We were unable to find a generic solution (cache invalidation is hard!), but thinking about this problem made me think of a much needed performance improvement: scheduling native updates.

This is pretty simple but will break even the current tests. When you set `textView.text = "a"` it natively runs `nativeView.setText("a");`, this is the same as you'd do in any app and usually the intended behavior. As we abstract more and more of the native API, a single change on the high level may call multiple changes on the same low level APIs, which have fixed costs.

Implementing a Microtask scheduler would solve most issues (this can probably be automated by the `Property` class):
```typescript
setText(text: string) {
    this.scheduleSetText(text);
}
setTextNative(...);
drainMicroTaskQueue() {
   // run all tasks for class
   this.queuedReferences = [];
   this.dirty = false;
}
scheduleSetText(text) {
   if(!this.dirty) { // native update not scheduled
      Promise.resolve().then(() => this.drainMicroTaskQueue());
   }
   // store text somewhere to avoid markingMode: none issues
   this.queuedReferences.push(text);
   this.tasks['textNativeUpdate'] = () => this.setTextNative(text);
   this.dirty = true;
}
```

But raises the issue:

```typescript
// old method
label.text // "ab"
span1.text = "c"; // triggers text change
label.text // "cb"
span2.text = "d"; // triggers text change
label.text // "cd"

// scheduler
label.text // "ab"
span1.text = "c";
label.text // "ab"
span2.text = "d";
label.text // "ab"
// end of macro task, triggers text change
setTimeout(() => label.text, // "cd"
    0);

// workaround for unit tests
label.text // "ab"
span1.text = "c";
label.applyNativeUpdates(); // calls drainMicrotaskQueue() under the hood
label.text // "cb"
span2.text = "d";
label.applyNativeUpdates();
label.text // "cd"
```

This kind of delay is already done when the view is Unloaded, for example: https://github.com/NativeScript/NativeScript/blob/864b51232b14a1b6add349f1a19659fa39f9a3a0/nativescript-core/ui/core/properties/properties.ts#L213

This is a big change that could begin as an optional flag on views, and may probably break some edge cases.

**Comment by farfromrefug** [comment]
@edusperoni I think we dont such a complicated behavior for now. This issue was more about the way spans are handled / updated, not the way we create the actual string from it.
Right now Span inherits `View` and that fact by itself makes them slow. Now in this issue in particular, one thing is that 6 spans will trigger 6 text updates for one color change on the label.
This can be easily solved (already done in my nativescript-htmllabel).

Now i also already handled in my plugin the fact that if during an update you change multiple properties which each will call an update on the text, it will be called multiple times. You can see it [here](https://github.com/Akylas/nativescript-label/blob/790f34f063c5a17568886893f6d855e8fbfb7232/src/label.ios.ts#L360) for ios

Finally to talk about the actual text creation i solved it multiple ways:
* do the whole string creations from spans fully natively: https://github.com/NativeScript/NativeScript/issues/8329. I already implemented it in `nativescript-htmllabel`
* create spannablestring from html. `nativescript-htmllabel` supports that. The actual creation of text from html is slower than doing it from spans. But seeing that i dont use span (and a lot of times  i am talking about 2 to 6 spans), html label is much faster in the whole app process.

**Comment by edusperoni** [comment]
The first part of my response was about the native creation itself, which is not the main reason for this issue (but impacts recyclerviews heavily, so I gave it as additional context), but the second part was the actual possible solution for the multiple updates in the core. I also believe that even if you move it to the native world you'll run into issues with big texts (especially with many spans). My scenario was the same as yours, I was trying to turn highly formatted HTML into a Label.

Span inheriting View is not the actual problem (although it sure is annoying), it's that every change on some span properties will emit an event that FormattedString receives and emits a new event which then TextBase receives and triggers a new update:

https://github.com/NativeScript/NativeScript/blob/master/nativescript-core/ui/text-base/formatted-string.ts#L133

https://github.com/NativeScript/NativeScript/blob/master/nativescript-core/ui/text-base/text-base-common.ts#L200

The only way to avoid this is delaying the updates until all the "current" code in the loop ran (by scheduing a microtask or timeout, for example). Looking at your code, I couldn't see what you did to prevent this issue without using a delay/microtask of sorts.

**Comment by farfromrefug** [comment]
Might point was not that what you say is not a good thing, but that it is not related to that specific issue. Actually here It is a bug ! Each span listen for the label style change and then trigger s a text update back on that label.
It is simply a bug not an issue with the core.
